builder/common/step_network_info.go

```diff
@@ -158,6 +161,66 @@ func (s *StepNetworkInfo) Run(ctx context.Context, state multistep.StateBag) mul
 		}
 	}
 
+	if s.AssociatePublicIpAddress != confighelper.TriUnset && s.SubnetId == "" {
```

Comment:
From looking at the code and original issue I understand the issue to occur when trying to set the value of AssociatePublicIPAddress to either true or false while using the default subnet and vpc, correct? 

If so the approach taken here is to get the ids for the default vpc and subnet if subnet_id is not provided before continuing so that the needed values are not empty.   Thus ensuring that s.SubnetId is never empty when we reach the condition `if s.SubnetId != "" && s.AssociatePublicIpAddress != confighelper.TriUnset {`. 

I got a little confused by the nested if/else statements. I think we can refactor this a bit given the information we already have by the time we get to this condition. For example we can add a guard statement and only continue if we need to. 

```go
	// By this point we know the Subnet, AZ, and VPC
	if s.AssociatePublicIpAddress == confighelper.TriUnset {
		state.Put("vpc_id", s.VpcId)
		state.Put("availability_zone", s.AvailabilityZone)
		state.Put("subnet_id", s.SubnetId)
		return multistep.ActionContinue
	}

	// Assuming Default VPC with default subnet code
	ui.Say(fmt.Sprintf("Setting public IP address to %t on instance running in the default subnet", *s.AssociatePublicIpAddress.ToBoolPointer()))

	if s.VpcId == "" {
```

There is a lot of boiler plate for the AWS describe calls. Nesting those make the code a little harder to read. Alternatively you might be able to take the approach of building default subnetFilterOptions and vpcFilterOptions early in the Run function if the following is true `s.SubnetId == "" && s.SubnetFilter.Empty()` or `s.VpcId == "" && s.VpcFilter.Empty()`